httpGet livenessprobe fails for controller and scheduler when following steps in scratch.md (fixes #12740)

aveshagarwal opened the pull request:
This PR updates scratch.md so that httpGet livenessprobe is sent to "host" : "127.0.0.1" for controller-manager and scheduler containers as they listen on 127.0.0.1 by default. If the "host" : "127.0.0.1" is not specified as part of the probe, the probe fails and both containers are restarted continuously. Another way to solve this is to start both containers with --address=0.0.0.0. This PR also updates the liveness userguide README to describe the host part of the httpGet probe.

Fixes #12740

aveshagarwal commented:
@pmorie

pmorie commented:
@aveshagarwal Actually, in this case, do we really need to specify host, since it will just use the pod IP if it isn't specificied?

aveshagarwal commented:
@pmorie As I understand, the following could be added to the liveness doc to describe http liveness probe:

http liveness probe performs an HTTP Get against the container's IP address on a specified port and path expecting on success that the response has a status code greater than or equal to 200 and less than 400 as follows:

 "livenessProbe": {
      "httpGet": {
        "host": "<>",
        "port": <>,
        "path": "<>"
      },
      "initialDelaySeconds":,
      "timeoutSeconds": 
 }

host specifies either the container's ip address, or if the container is sharing network with host but is listening on 127.0.0.1, then host should be specified as 127.0.0.1.

aveshagarwal commented:
@pmorie updated the PR.

aveshagarwal commented:
@pmorie I went through the steps in scratch.md and I was able to reproduce the issue 12740. I had to specify "host": "127.0.0.1" to pass livenessprobe otherwise controller manager and scheduler containers kept restarting by kubelet.

aveshagarwal commented:
#netstat -tulnp|grep kube
tcp        0      0 127.0.0.1:10248         0.0.0.0:*               LISTEN      17018/kubelet       
tcp        0      0 192.168.122.111:10250   0.0.0.0:*               LISTEN      17018/kubelet       
tcp        0      0 127.0.0.1:10251         0.0.0.0:*               LISTEN      17360/kube-schedule 
tcp        0      0 127.0.0.1:10252         0.0.0.0:*               LISTEN      17334/kube-controll 
tcp        0      0 192.168.122.111:10255   0.0.0.0:*               LISTEN      17018/kubelet       
tcp        0      0 192.168.122.111:8080    0.0.0.0:*               LISTEN      17371/kube-apiserve 
tcp6       0      0 :::4194                 :::*                    LISTEN      17018/kubelet       
tcp6       0      0 :::6443                 :::*                    LISTEN      17371/kube-apiserve 
#netstat -tulnp|grep etcd
tcp6       0      0 :::4001                 :::*                    LISTEN      17293/etcd          
tcp6       0      0 :::2379                 :::*                    LISTEN      17293/etcd          
tcp6       0      0 :::2380                 :::*                    LISTEN      17293/etcd          
tcp6       0      0 :::7001                 :::*                    LISTEN      17293/etcd

aveshagarwal commented:
Without "host" : "127.0.0.1", kubelet container shows following probe failures:

I0827 09:11:36.729154       1 manager.go:1446] pod "kube-controller-manager-192.168.122.111_default" container "kube-controller-manager" is unhealthy (probe result: failure), it will be killed and re-created.
I0827 09:11:36.730350       1 manager.go:1446] pod "kube-scheduler-192.168.122.111_default" container "kube-scheduler" is unhealthy (probe result: failure), it will be killed and re-created.

aveshagarwal commented:
@pmorie, or another way is to solve it to start controller-manager and scheduler with --address=0.0.0.0 if we do not want to specify "host" : "127.0.0.1" in httpget livenessprobe.

brendandburns commented on docs/user-guide/liveness/README.md:
````diff
@@ -69,7 +69,7 @@ The [http-liveness.yaml](http-liveness.yaml) demonstrates the HTTP check.
 ```
 
-The Kubelet sends a HTTP request to the specified path and port to perform the health check. If you take a look at image/server.go, you will see the server starts to respond with an error code 500 after 10 seconds, so the check fails.
+The Kubelet sends a HTTP request to the specified path and port to perform the health check. If you take a look at image/server.go, you will see the server starts to respond with an error code 500 after 10 seconds, so the check fails. The Kubelet sends the probe to the container's ip address by default which could be specified with `host` as part of httpGet probe. If the container listens on `127 ...
````
"sends an HTTP request"

brendandburns commented:
Yes, this is correct.  Added one minor nit...

@k8s-bot test this please (since there no way it could break e2e)

Commit:
Updates scratch.md to send httpGet livenessprobe to 127.0.0.1 for
controller-manager and scheduler as they both listen on 127.0.0.1 by
default. This fixes #12740.

Commit:
Updates the liveness user guide README to describe the host part of
httpGet probe.

aveshagarwal commented:
@brendandburns Fixed and pushed updated PR.

brendandburns commented:
LGTM, closing and re-opening to get shippable to run...

Fleshgrinder commented on docs/user-guide/liveness/README.md:
````diff
@@ -69,7 +69,7 @@ The [http-liveness.yaml](http-liveness.yaml) demonstrates the HTTP check.
 ```
 
-The Kubelet sends a HTTP request to the specified path and port to perform the health check. If you take a look at image/server.go, you will see the server starts to respond with an error code 500 after 10 seconds, so the check fails.
+The Kubelet sends an HTTP request to the specified path and port to perform the health check. If you take a look at image/server.go, you will see the server starts to respond with an error code 500 after 10 seconds, so the check fails. The Kubelet sends the probe to the container's ip address by default which could be specified with `host` as part of httpGet probe. If the container listens on `12 ...
````
You boys really need to start wrapping at 80 to support reviewing since reviewing is line based. 😉
